app/views/AuthenticationWebView.js

```diff
@@ -58,37 +62,58 @@ class OAuthView extends React.PureComponent {
 		this.setState({ logging: true });
 
 		try {
-			await RocketChat.loginOAuth(params);
+			await RocketChat.loginOAuthOrSso(params);
 		} catch (e) {
 			console.warn(e);
 		}
 		this.setState({ logging: false });
 		this.dismiss();
 	}
 
+	onNavigationStateChange = (webViewState) => {
+		const url = decodeURIComponent(webViewState.url);
+		if (this.authType === 'saml' || this.authType === 'cas') {
+			const { navigation } = this.props;
+			const ssoToken = navigation.getParam('ssoToken');
+			if (url.includes('ticket') || url.includes('validate')) {
+				let payload;
+				if (this.authType === 'saml') {
+					payload = `{ "saml": true, "credentialToken": "${ ssoToken }" }`;
+				} else {
+					payload = `{ "cas": { "credentialToken": "${ ssoToken }" } }`;
+				}
+				// We need to set a timeout when the login is done with SSO in order to make it work on our side.
+				// It is actually due to the SSO server processing the response.
+				setTimeout(() => {
+					this.login(JSON.parse(payload));
+				}, 3000);
```

Comment:
Does it needs to be whole 3 seconds?
What about a poor internet connection?
Would it work properly?

Reply:
Yes, unfortunately it need to be as it is - as explained on the comment. It is not an issue from our side (read the comment pls).

We've done lot's of tests when we had implemented the feature on the native apps. Refer to the delay [here](https://github.com/RocketChat/Rocket.Chat.Android/blob/e9821393c593bac04856b6bde4b3e03428362c68/app/src/main/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenter.kt#L113) for the Android Native and [here](https://github.com/RocketChat/Rocket.Chat.iOS/blob/42b9857ef445a238fb915716a68bbdd5210b38a6/Rocket.Chat/Controllers/Auth/CASViewController.swift#L76) for the iOS Native.

About the question with a poor internet connection, then I advise implementing a kind of retry IO like we have done on native, but then it needs to be done not only for CAS, SAML, but for all supported authentication we have. This way we can ensure a kind of repeated callback to handle this kind of situation.